### ISSUE TITLE
A11y: Decorative images in a figure element cause VoiceOver screen reader reading the file name.

### ISSUE THREAD

**Issue by juliemoynatPro** [WordPress/gutenberg#75027]
### Description

When you add an image in a post, it is put inside a `figure` HTML element, which is a semantic HTML element.

So, when the image is decorative, it should not be inside a `figure` but it is.

So, VoiceOver, the screen reader for MacOS and iOS, does not ignore these images because the `figure` tells that they're not decorative images. And VoiceOver reads the filename for the images because the `alt` attribute is empty (`alt=""`).

In order to fix it:
- Recommended: Only images with a caption (`figcaption` element) should be inside a `figure` element;
- Or, you can eventually add an `aria-hidden="true"` attribute on `figure` only when the image has no alternative text.

### Step-by-step reproduction instructions

1. Add an image block in a post with an image without alternative text;
2. See the code (an `img` element with empty `alt` attribute (`alt=""`) inside a `figure` element);
3. Make VoiceOver (on MacOS or iOS) read the image: it will read the filename;
4. Move the image outside of the `figure` (with the browser inspector) and VoiceOver will ignore the image as it should do.

### Screenshots, screen recording, code snippet

Example of generated code to understand:

```html
<figure class="wp-block-image size-large">
  <img loading="lazy" decoding="async" width="1024" height="573" data-id="136327" src="https://example.com/wp-content/uploads/2025/01/image-1024x573.jpg.webp" alt="" class="wp-image-136327" srcset="https://example.com/wp-content/uploads/2025/01/image-1024x573.jpg.webp 1024w, https://example.com/wp-content/uploads/2025/01/image-300x168.jpg.webp 300w, https://example.com/wp-content/uploads/2025/01/image-150x84.jpg.webp 150w, https://example.com/wp-content/uploads/2025/01/image-768x430.jpg.webp 768w, https://example.com/wp-content/uploads/2025/01/image.jpg.webp 1200w" sizes="auto, (max-width: 1024px) 100vw, 1024px">
</figure>
```

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [x] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

**Comment by Rishit30G** [comment]
Thanks for sharing the issue, 

However I wasn't able to replicate properly from Step 3, could you possibly share a screencast that should help me understand this better 🙇🏻 

When I tried replicating the issue, for me even without the alt tag the image was ignored when using VoiceOver

https://github.com/user-attachments/assets/a04b6fd0-11ac-41f8-a7ce-0ed72ea57eb6

**Comment by dpmehta** [comment]
I have also tried to reproduce the issue but not able to , the image is already being skipped by VoiceOver even if it doesn't have alt text empty.

https://github.com/user-attachments/assets/63465032-e705-4321-ae54-a149b8065078

**Comment by SirLouen** [comment]
Hello @juliemoynatPro it seems that there have been a couple intents to reproduce unsucessfully. Could you provide more info or check the screencasts provided to see what they are not doing according to your instructions?

**Comment by joedolson** [comment]
Semantically, @juliemoynatPro is right; the `figure` element should only be used for content - by definition, a decorative image is not content.

That said, I also could not reproduce this experience in VoiceOver/MacOS. I tested in normal reading and in graphic navigation. Reading ignored the image and graphic navigation believed there were no images on the page.

Can you specify what version of which OSes you were testing in, @juliemoynatPro?

**Comment by juliemoynatPro** [comment]
I'm sorry, I totally forgot to tell that I'm testing on Safari and it seems that there is not exactly the same behaviour on Chrome.

I'm testing with:
- VoiceOver with MacOS Tahoe 26.2 + Safari 26.2
- VoiceOver with iOS 26.2.1 (iPhone) + Safari

On the WordPress webpage I'm testing, VoiceOver on Chrome says "number" for each decorative image inside a figure but I can't reproduce this on [the CodePen I've made to test outside WordPress](https://codepen.io/copsae/pen/jErxWoK) so I don't know what's the problem.

Anyway, as you can see in this screenshot of my CodePen, VoiceOver is telling the name of the file "960px-Mesang099d.jpg, image" (sorry, I can't make a video but it really says this out loud).

<img width="631" height="600" alt="Screenshot of the CodePen result with what VoiceOver is saying." src="https://github.com/user-attachments/assets/35b1bd4d-b153-47bb-8c38-6c49f84018ef" />

**Comment by dpmehta** [comment]
As per @juliemoynatPro suggestion i tried to reproduce the issue in safari browser , i am able to reproduce the issue , VoiceOver not skipping decorative image instead telling name of the image , this is strange in chrome VoiceOver skips image successfully as intended , Please share your thoughts about this

https://github.com/user-attachments/assets/aed4ff50-9f55-4dff-b12e-502d903ccc41

**Comment by juliemoynatPro** [comment]
I said:

> On the WordPress webpage I'm testing, VoiceOver on Chrome says "number" for each decorative image inside a figure but I can't reproduce this on [the CodePen I've made to test outside WordPress](https://codepen.io/copsae/pen/jErxWoK) so I don't know what's the problem.

I know what's going on: VoiceOver doesn't say "number" but "nombre" in French (because my Mac is in French) and the correct translation into English I should have done is "figure".

Actually, my images are in a Gallery block so they are each in a `figure` and both `figure` are inside a `figure` to group them (why?). So, VoiceOver on Chrome says "figure" for each decorative image.

But, I've added a similar code in my CodePen and I can't reproduce. I've tested on the WordPress playground and I can't reproduce. I don't know why it can say "figure" in some cases and not others… That's really strange.

What's sure is that the use of `figure` everywhere even for non-content element is not correct.

**Comment by joedolson** [comment]
I tested with NVDA on Windows with multiple browsers and on Safari/VoiceOver on MacOS, both in the CodePen provided and in a page I created in WordPress with comparable content. But I couldn't get the figure to be announced in any of those cases. I had VoiceOver's verbosity set to High with no customizations. 

It seems plausible that there is some case where the existence of a `figure` is announced regardless of it not having any naming content, but as long as we can't reproduce it, it's pretty hard to know what that case is and whether we can do anything about it. 

I do think that not using `figure` when there's no naming context would be beneficial. It may be a hard change to get made, because there are a lot of styles specifically tied to `figure` in themes, and the change would be likely to break styling on some number of sites because of that. I'm not sure it would be fully justified to make the change for semantic purity, if there's no real-world example of it causing a problem.

This would be that example, if we can reliably reproduce the scenario.

**Comment by juliemoynatPro** [comment]
@joedolson I don't understand because @dpmehta reproduced the problem of the file name read by VoiceOver on Safari. So the first problem I've submitted is reproduced. It's not semantic purity because the problem of file name reading by the screen reader for decorative images is real.

~~I thought of a more simple solution: adding a `role="presentation"` on each `figure` element that have no `figcaption` inside. This would have no incidence on the styles made by themes but it would remove semantic from `figure` that should not be there so they will just behave as `div`.~~

Edit: Adding a `role="presentation"` on the `figure` element doesn't change anything. It doesn't work (even with `role="none"`). So **the only solution I see is: adding an `aria-hidden="true"` attribute on the `figure` element if the image has an empty alt _and_ if there is no `figcaption`.**

What do you think?

**Comment by joedolson** [comment]
@juliemoynatPro Yes, @dpmehta was able to personally reproduce the issue. However, what we need to be able to move forward is to know *how* to reproduce it. The fact that a couple of instances are able to reproduce the issue doesn't help when we also have a couple of instances that fail to reproduce it. We need to know what the differences are that result in this; otherwise, we can't take meaningful action on it.

What I want to know is whether this is unique to a specific version, to a specific collection of settings, etc. That would allow me to do some research to learn why it's happening. 

I'm a little uncomfortable with adding `aria-hidden="true"` to a `figure`, but from a practical standpoint, if this is only happening when the contained image has no alt text & no caption, then it's not super harmful. It would also need to know whether the image has a link, to avoid creating a silent tab stop. (Not that the experience would be a good one, but, we're talking about pretty poor situations here.)

**Comment by juliemoynatPro** [comment]
@joedolson I'm using VoiceOver on MacOS and iOS (iPhone) with default settings. I use these devices only for accessibility testing so there is nothing special.

An accessibility friend tested too and reproduced the problem on my CodePen and on a WordPress website with: iOS 26.0 + Safari, and macOS 14 + Safari 26.1.

**Comment by joedolson** [comment]
Thanks. It may be something specific to more recent versions; I was testing with Safari 18 and iOS 18. So it sounds like something has changed in how Apple interprets unnamed `figure` elements.

This is non-trivial to fix, unfortunately. We can make a change in how future images are rendered, but there's little we can do about past image content - they'll just continue to be wrong. 

@jeryj @t-hamano Per the above conversation, `figure` elements that contain images without alt attributes or captions are now being announced, where they used to be ignored. This means we need to handle this case differently in block content - either of you have any bandwidth to work on this?

**Comment by dpmehta** [comment]
I used macOS Tahoe version 26.2 and Safari with version 26.2,With this configuration i am able to reproduce the issue as i shared in video above, I also align with fix, if image does not have alt text and caption is empty then that image should be sklipped ( may be we can add aria-hidden as true for such images).

**Comment by t-hamano** [comment]
Thanks for the ping.

> This means we need to handle this case differently in block content

What would be the most ideal approach from an accessibility standpoint? Would just adding `aria-hidden="true"` be enough?

One approach might be to use a `div` tag instead of a `figure` tag for images that don't have alt or captions, but this may affect many themes. There are likely some themes that use the figure element as part of their selector.

**Comment by joedolson** [comment]
@t-hamano I think that changing the element is a non-starter, for that reason. There are definitely themes and plugins that use the `figure` element as a selector, and it's hard to imagine that working out well.

Adding `aria-hidden="true"` is enough, as long as we impose enough conditions on it - it needs to only present if there is no naming content. Empty `alt` text, no caption, etc, and it needs to be removed if that information is added.

From an accessibility standpoint, I'd prefer to remove the figure; I just don't think that's a viable approach for back compat reasons.

**Comment by t-hamano** [comment]
I re-tested this issue. Here is my test environment.

- OS: MacOS Tahoe 26.4.1
- Browser: 26.4
- Screen Reader: VoiceOver (Quick Navigation Mode)

I used [this Codepen](ttps://codepen.io/editor/Tetsuaki-Hamno/pen/019dc32c-22ba-7bef-a34e-ed5c5b91f352) to see how VoiceOver reads aloud when an image is focused using the down arrow key.

- One image in a figure: `frame {file-name} image`
- One image in a figure and aria-hidden="true": `frame image`
- One image in a div: `frame image`
- One image in a div and aria-hidden="true": `frame image`

https://github.com/user-attachments/assets/d2c21e15-2a23-49cf-9591-f08660167133

My tests show that adding `aria-hidden=true` to a `figure` element can change the screen reader's reading from `frame {file-name} image` to `frame image`, but it doesn't prevent Safari from reading the image itself. Is my understanding correct?

**Comment by joedolson** [comment]
Have you tried to see what it does if the image has `role="none"`? 

`aria-hidden` *should* remove the element and all its children from the accessibility tree, so it seems to me like Safari is just violating that expectation. This isn't surprising; Safari isn't particularly good about respecting the standards. 

I'm wondering if we can apply https://github.com/WordPress/gutenberg/pull/78064, and use that to resolve this by marking the image as decorative using `role`.

**Comment by t-hamano** [comment]
Now that #78064 has been merged, I believe we can close this issue.

**Comment by juliemoynatPro** [comment]
The `role="none"` on the image fixes the problem but the PR #78064 will not be retroactive: people will need to check the checkbox for all their decorative images even if they have 1000 posts with several decorative images inside…

**Comment by t-hamano** [comment]
> people will need to check the checkbox for all their decorative images even if they have 1000 posts with several decorative images inside…

I understand this, of course. However, it is practically impossible to automatically determine whether an image already present in the content is decorative. Even without captions or alt text, a user might want to represent it as a non-decorative image. Converting existing images to decorative ones based on some rules could lead to significant problems.

**Comment by juliemoynatPro** [comment]
I don't understand. It doesn't make sense.

All websites using WordPress with Gutenberg and having images with no alternative text nor caption in their pages and posts have an accessibility bug because VoiceOver with Safari reads the file name of images.

Whether the images are supposed to be decorative or not is not the problem here. If the images are supposed to be informative and need an alternative text so, the file name of the image should not be read: WordPress users need to fill the alternative text. That's all.

So if we just content ourselves with that PR #78064, the accessibility problem, created by Gutenberg adding `figure` around each images even if they are decorative and have no caption, will remain on all those websites because no one will do the fix on all their pages/posts and images and **no one will know that they should fix it by manually checking that checkbox**.

Gutenberg have created an accessibility bug so Gutenberg have to fix it.

**Comment by joedolson** [comment]
Based on this issue, this appears to be a bug in Safari/VoiceOver, not in WordPress. The code that WordPress creates meets standards and works in the majority of cases, but Safari/Voiceover doesn't handle that case as expected.

While WordPress can take steps to work around browser bugs like this, making automatic edits to users content is not something that we would take on lightly. It raises many additional complicated questions, such as

 - Should these changes introduce additional revisions in content, increasing database usage for all users by a value equal to their current size of drafted posts? A revision would allow users to roll back if there were any problems, but may be disabled on some sites.
 - Should these changes alter post modified dates? This would be necessary in order to create revisions, but would also make it impossible for users to identify when they last *actually* modified their content.
 - Is this a one-time thing, or does it need to be modified on all content, forever? If plugins insert content using this pattern in blocks, editing the content could break the block. Does this mean the change actually needs to happen on the front-end? 
 - What if the content actually doesn't contain the alt text in the admin, because the alternative text is inserted dynamically on the front-end? How would we know that?

It's not reasonable or practical for WordPress to make a change to user's content of this scale. It is unfortunate that this bug in Safari has introduced an accessibility problem in WordPress content, but it would be *vastly* more effective and simpler to address this bug on the Safari side than on the WordPress side.
